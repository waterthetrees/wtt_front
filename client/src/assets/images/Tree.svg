<?xml version="1.0" encoding="UTF-8"?><svg id="Happy_Color" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><defs><style>.cls-1,.cls-2{fill:none;stroke:#000;stroke-linecap:round;stroke-linejoin:round;}.cls-2{stroke-width:2px;}</style></defs><line class="cls-1" x1="12" y1="23.98" x2="12" y2="7.35"/><line class="cls-1" x1="12" y1="10.94" x2="15.28" y2="9.54"/><line class="cls-1" x1="12" y1="14.11" x2="8.62" y2="12.47"/><path class="cls-2" d="M12,16.37c-2.8,2.65-7.35,1.95-9.22-1.42-1.87-3.37-.06-7.6,3.67-8.58,.13-2.97,2.58-5.32,5.55-5.32s5.43,2.35,5.55,5.32c3.73,.98,5.54,5.21,3.67,8.58-1.87,3.37-6.42,4.07-9.22,1.42Z"/></svg>