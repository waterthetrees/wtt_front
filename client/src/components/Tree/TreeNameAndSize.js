import React, { useEffect, useState } from 'react';
import { useFormContext } from 'react-hook-form';
import { InputAdornment } from '@mui/material';
import { FormAutocomplete, FormDecimalField } from '../Form';
import { trees } from '../../data/dist/trees';
import sortTreesBy from '../../data/sortTreesBy';

function getNames(treeList, name, firstItem) {
  // Add the names to a Set and then expand it to an array to ensure the lists of names are unique.
  // Otherwise, React will complain about duplicate keys in the menu.  Sort the trees by the
  // different fields before extracting the strings, so that the sort function can access the
  // per-tree lowercase sort fields.
  return firstItem.concat([...new Set(treeList.sort(sortTreesBy[name]).map((tree) => tree[name]))]);
}

// Adjust the adornment label so it aligns with the input placeholder.
const unit = (label) => ({
  endAdornment: <InputAdornment position="end" sx={{ marginBottom: '2px' }}>{label}</InputAdornment>,
});

const vacantLabel = 'Vacant Site';

export default function TreeNameAndSize() {
  const { watch, setValue } = useFormContext();
  // Get the value of common that was specified by our parent's defaultValues, so that we can
  // default the lastCommon state to it.
  let [common, scientific, genus] = watch(['common', 'scientific', 'genus']);
  const [lastCommon, setLastCommon] = useState(common);

  useEffect(() => {
    if (common !== lastCommon) {
      const commonTrees = trees.filter((tree) => tree.common === common);

      if (commonTrees.length) {
        // Default scientific and genus to the first matching tree (in most cases, the only one).
        ({ scientific, genus } = commonTrees[0]);
      } else if (common === vacantLabel) {
        scientific = genus = vacantLabel;
      } else if (common === null) {
        // When a field is cleared, it gets set to null, but calling setValue() with null doesn't
        // seem to update the other fields.  Setting them to '' works for some reason.
        scientific = genus = '';
      }

      setValue('scientific', scientific);
      setValue('genus', genus);
      setLastCommon(common);
    }
  }, [setValue, common, lastCommon]);

  const allFieldsEmpty = !common && !scientific && !genus;
  const firstItem = allFieldsEmpty || [common, scientific, genus].includes(vacantLabel)
    ? [vacantLabel]
    : [];
  const treeList = allFieldsEmpty
    ? trees
    // Find the union of trees that match one or more of the names.
    : trees.filter((tree) => (tree.common === common
      || tree.scientific === scientific
      || tree.genus === genus));
  const commonNames = getNames(treeList, 'common', firstItem);
  const scientificNames = getNames(treeList, 'scientific', firstItem);
  const genusNames = getNames(treeList, 'genus', firstItem);

  return (
    <>
      <FormAutocomplete
        name="common"
        label="Common name"
        options={commonNames}
        rules={{ required: true, minLength: 1, maxLength: 100 }}
      />

      <FormAutocomplete
        name="scientific"
        label="Scientific name"
        options={scientificNames}
        rules={{ minLength: 1, maxLength: 100 }}
      />

      <FormAutocomplete
        name="genus"
        label="Genus"
        options={genusNames}
        rules={{ minLength: 1, maxLength: 100 }}
      />

      <FormDecimalField
        name="height"
        label="Height"
        InputProps={unit('feet')}
      />

      <FormDecimalField
        name="dbh"
        label="Diameter"
        title="Diameter at breast height"
        InputProps={unit('inches')}
      />
    </>
  );
}
