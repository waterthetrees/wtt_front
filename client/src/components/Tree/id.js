const MIN_LAT = -90;
const MAX_LAT = 90;
const MIN_LON = -180;
const MAX_LON = 180;

export const hashcode = (s) => {
  let h = 0;
  let i = s.length;
  while (i > 0) {
    h = ((h << 5) - h + s.charCodeAt(--i)) | 0;
  }
  return Math.abs(h);
};

// This code originally appeared here: https://github.com/sunng87/node-geohash/blob/master/main.js#L127-L161
export const geohashToInt = (latitude, longitude, bitDepth) => {
  bitDepth = bitDepth || 52;

  let bitsTotal = 0;
  let combinedBits = 0;
  let mid = null;
  let maxLat = MAX_LAT;
  let minLat = MIN_LAT;
  let maxLon = MAX_LON;
  let minLon = MIN_LON;

  while (bitsTotal < bitDepth) {
    combinedBits *= 2;
    if (bitsTotal % 2 === 0) {
      mid = (maxLon + minLon) / 2;
      if (longitude > mid) {
        combinedBits += 1;
        minLon = mid;
      } else {
        maxLon = mid;
      }
    } else {
      mid = (maxLat + minLat) / 2;
      if (latitude > mid) {
        combinedBits += 1;
        minLat = mid;
      } else {
        maxLat = mid;
      }
    }
    bitsTotal++;
  }
  return combinedBits;
};

export const IDForTree = (data) => {
  const [common, scientific, city, lat, lng] = data;
  const hashed = geohashToInt(lat, lng, 52);
  const encoded = hashcode(`${city}-${common}-${scientific}-${hashed}`);
  return encoded;
};

export const GetId = (data) => {
  if (data.filter(Boolean).length === 5) {
    return IDForTree(data);
  }
  return null;
};
