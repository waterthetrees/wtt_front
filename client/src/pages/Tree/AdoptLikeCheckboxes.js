import React, { useState } from 'react';
import { useAuth0 } from '@auth0/auth0-react';
import { Box, Grid, Link } from '@mui/material';
import format from 'date-fns/format';
import {
  useTreeAdoptionsQuery, useTreeLikesQuery, useTreeAdoptionsMutation,
  useTreeHistoryMutation, useTreeLikesMutation,
  useCreateTreeDataMutation,
} from '@/api/queries';
import { AdoptionCheckbox, InfoCheckbox, StarCheckbox } from '@/components/Checkbox';
import { TooltipTop } from '@/components/Tooltip';
import AdoptionDirections from '@/pages/Adopt/AdoptionDirections';
import { useDebouncedCallback } from 'use-debounce';

function Liked({ handleChange, currentTreeId, user }) {
  const { data: { liked, likedCount } } = useTreeLikesQuery({
    id: currentTreeId,
    email: user.email,
  });

  return (
    <Grid item sx={{ mr: 1 }}>
      <Grid container alignItems="center">
        <TooltipTop title="Like">
          <StarCheckbox
            edge="start"
            name="liked"
            checked={liked}
            onChange={handleChange}
          />
        </TooltipTop>
        <Box fontSize="1.125rem">{likedCount}</Box>
      </Grid>
    </Grid>
  );
}

function Adopted({
  handleChange, user, currentTreeId, directions, showDirections,
}) {
  const { data: { adopted, adoptedCount } } = useTreeAdoptionsQuery({
    id: currentTreeId,
    email: user.email,
  });

  return (
    <Grid item>
      <Grid container alignItems="center">
        <TooltipTop title="Adopt">
          <AdoptionCheckbox
            name="adopted"
            checked={adopted}
            onChange={handleChange}
          />
        </TooltipTop>
        <Box sx={{ fontSize: '1.125rem', mr: 1 }}>{adoptedCount}</Box>
        <InfoCheckbox
          checked={directions}
          onClick={() => showDirections(!directions)}
        />
      </Grid>
    </Grid>
  );
}

export default function AdoptLikeCheckboxes({
  currentTreeData, edit, isTreeQueryError,
}) {
  const { user } = useAuth0();
  const mutateTreeLikes = useTreeLikesMutation();
  const mutateTreeAdoptions = useTreeAdoptionsMutation();
  const mutateHistory = useTreeHistoryMutation();
  const mutateCreateTreeData = useCreateTreeDataMutation();
  const [directions, showDirections] = useState(false);

  const handleChange = (event) => {
    if (isTreeQueryError) {
      mutateCreateTreeData.mutate({
        ...currentTreeData,
        scientific: currentTreeData.scientific
        || currentTreeData.genus,
        city: currentTreeData.city || currentTreeData.sourceId,
        volunteer: user.nickname,
        health: currentTreeData.health || 'fair',
      });
    }

    const sendTreeHistory = {
      id: currentTreeData.id,
      dateVisit: format(new Date(), 'yyyy-MM-dd HH:mm:ss'),
      [event.target.name]: event.target.checked,
      volunteer: user.nickname,
    };
    const sendTreeUser = {
      id: currentTreeData.id,
      common: currentTreeData.common,
      nickname: user.nickname,
      email: user.email,
      request: {
        name: event.target.name,
        type: event.target.checked ? 'POST' : 'DELETE',
      },
    };

    if (event.target.name === 'adopted') {
      setTimeout(() => {
        mutateTreeAdoptions.mutate(sendTreeUser);
        mutateHistory.mutate(sendTreeHistory);
      }, 500);
      showDirections(event.target.checked);
    }
    if (event.target.name === 'liked') {
      setTimeout(() => {
        mutateTreeLikes.mutate(sendTreeUser);
        mutateHistory.mutate(sendTreeHistory);
      }, 500);
    }
  };

  if (!user) return null;

  return (
    <>
      <Grid container alignItems="center" justifyContent="space-between" flexWrap="initial">
        <Grid container>
          <Liked
            currentTreeId={currentTreeData.id}
            user={user}
            handleChange={handleChange}
          />
          <Adopted
            currentTreeId={currentTreeData.id}
            user={user}
            directions={directions}
            showDirections={showDirections}
            handleChange={handleChange}
          />
        </Grid>
      </Grid>
      {directions && (
        <AdoptionDirections
          open={directions}
          onClose={() => showDirections(false)}
        />
      )}
    </>
  );
}
