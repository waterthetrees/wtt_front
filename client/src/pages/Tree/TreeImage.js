import React, { useEffect, useState } from 'react';
import treeImages from '@/data/dist/treeImages.json';

export const ImageLoad = React.memo(({ src, placeholder, alt = '' }) => {
  const [loading, setLoading] = useState(true);
  const [currentSrc, updateSrc] = useState(placeholder);

  useEffect(() => {
    if (src !== currentSrc) updateSrc();
    // start loading original image
    const imageToLoad = new Image();
    imageToLoad.src = src;
    imageToLoad.onload = () => {
      // When image is loaded replace the src and set loading to false
      setLoading(false);
      updateSrc(src);
    };
  }, [src]);

  return (
    <img
      src={currentSrc}
      style={{
        opacity: loading ? 0.1 : 1,
        transition: 'opacity .30s linear',
        width: '100%',
      }}
      placeholder={placeholder}
      alt={alt}
      loading="lazy"
    />
  );
});
ImageLoad.displayName = 'ImageLoad';

export const fixScientificForImage = (scientific) => {
  if (!scientific || !treeImages[scientific]) return null;

  // replace spaces with hyphens for image path using rejex
  const scientificNospaces = scientific.replace(/\s/g, '-');

  const imagePath =
    `../../assets/images/data/${scientificNospaces}.jpg` ||
    treeImages[scientific]?.imageURL;
  return imagePath;
};
