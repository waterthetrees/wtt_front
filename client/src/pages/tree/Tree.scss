@import "../../styles/app";

.MapContainer {
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  height: 100vh;
  width: 100%;
  margin: -100px;
}

.tree__label {
  color: #ff0000;
}

.tree__label img {
  height: 50px;
}

.tree__add {
  position: absolute;
  bottom: 10px;
  left: 10px;
  z-index: 1049;
  border-radius: 50%;
  padding: 0;
  .btn {
    border: 0px solid #000;
    border-radius: 50%;
  }
  .btn-success:active {
    color: #fff;
    background-color: #006400;
  }

  h1,
  .h1 {
    font-size: $font-size-xl;
    font-weight: $font-weight-bold;
    margin-bottom: 0;
    font-weight: 900;
    line-height: 1;
  }
}

.tree__add:not(:disabled):not(.disabled).active,
.tree__add:not(:disabled):not(.disabled):active,
.show > .tree__add.dropdown-toggle {
  color: #fff;
  background-color: #006400;
  border-color: white;
}
.tree__add .active {
  color: #fff;
  background-color: #006400 !important;
  border-color: white;
}

.tree__add:active {
  color: #fff;
  background-color: #006400;
  border-color: white;
}

.tree__add .active:hover {
  color: #fff;
  background-color: #006400;
  border-color: white;
}

.tree__add:hover {
  color: #006400;
  background-color: white;
  border-color: white;
}

.userprofile {
  position: absolute;
  right: 10px;
  bottom: 10px;
  width: 50px;
  height: 50px;
  z-index: 1049;
  border-radius: 50%;
  padding: 0;
  .btn {
    border: 0px solid #000;
    border-radius: 50%;
  }
  .btn-success:active {
    color: #fff;
    background-color: #006400;
  }

  h1,
  .h1 {
    font-size: $font-size-xl;
    font-weight: $font-weight-bold;
    margin-bottom: 0;
    font-weight: 900;
    line-height: 1;
  }
}
