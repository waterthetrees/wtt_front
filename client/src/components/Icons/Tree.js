import React from 'react';
import TreeImage from '@/assets/images/addtree/treefattrunk.png';
import { CardMedia } from '@mui/material';
import { styled } from '@mui/material/styles';

export const TreeLogo = () => {
  const StyledLogo = styled(CardMedia)({
    width: '32px',
    aspectRatio: '1/1',
  });

  return <StyledLogo component="img" src={TreeImage} alt="tree"></StyledLogo>;
};
