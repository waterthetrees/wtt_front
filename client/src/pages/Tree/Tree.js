import React from 'react';
import { Alert } from '@mui/material';
import TreeHeader from './TreeHeader';
import TreeMaintenance from './TreeMaintenance';
import TreeRemoval from './TreeRemoval';
import TreeHealth from './TreeHealth';
import TreeNotes from './TreeNotes';
import TreeHistory from './TreeHistory';
import TreeInfo from './TreeInfo';
import TreeLinks from './TreeLinks';
import { ImageLoad } from './TreeImage';

import treeImages from '@/data/dist/treeImages.json';
import { CarbonCalculator } from './CarbonCalculator';

const undefRequiredField = (requiredField) =>
  typeof requiredField === 'undefined';

export const checkForUnfitData = (currentTreeData) =>
  [currentTreeData?.common, currentTreeData?.scientific].some(
    undefRequiredField,
  ) ||
  [
    'vacant',
    'vacant site',
    'unsuitable site',
    'asphalted well',
    'planting site',
    'other',
    '#n/a',
    '::',
  ].includes(currentTreeData?.common.toLowerCase());
// Common name field has inconsistent data when a tree's missing.
// Until we clean up the data on the way into the vector tiles,
// We'll add it here so that Maintenance/Health/removal don't render.

export default function Tree({
  TreeDetailsContainer,
  drawerWidth,
  currentTreeData,
  currentTreeId,
  setCurrentTreeId,
  isTreeQueryError,
}) {
  const { scientific } = currentTreeData || {};
  // If a tree is selected but there was an error in fetching the data, show an error message.
  // Otherwise, show a blank panel while waiting for the data.
  const noDataChild =
    currentTreeId && isTreeQueryError ? (
      <Alert severity="error" sx={{ fontSize: '1.2rem' }}>
        No data is available for this tree.
      </Alert>
    ) : null;

  // check for malformed, missing common, scientific names, vacant sites, etc.
  const hasUnfitData = checkForUnfitData(currentTreeData);

  const handleClose = () => setCurrentTreeId(null);

  return (
    <TreeDetailsContainer
      title="Tree Details"
      width={drawerWidth}
      open={!!currentTreeId}
      onClose={handleClose}
    >
      {currentTreeData ? (
        <>
          <TreeHeader
            currentTreeData={currentTreeData}
            hasUnfitData={hasUnfitData}
            isTreeQueryError={isTreeQueryError}
          />

          {scientific &&
            Object.prototype.hasOwnProperty.call(treeImages, scientific) && (
              <ImageLoad
                src={treeImages[scientific]}
                placeholder="placeholder.jpg"
              />
            )}

          {!hasUnfitData && (
            <TreeHealth
              currentTreeData={currentTreeData}
              isTreeQueryError={isTreeQueryError}
            />
          )}

          {!hasUnfitData && (
            <TreeNotes
              currentTreeData={currentTreeData}
              isTreeQueryError={isTreeQueryError}
            />
          )}

          {!hasUnfitData && (
            <TreeMaintenance
              currentTreeData={currentTreeData}
              isTreeQueryError={isTreeQueryError}
            />
          )}

          {currentTreeId && <TreeHistory currentTreeId={currentTreeId} />}

          <TreeInfo currentTreeData={currentTreeData} />

          <TreeLinks currentTreeData={currentTreeData} />

          <CarbonCalculator currentTreeData={currentTreeData} />

          {!hasUnfitData && (
            <TreeRemoval
              currentTreeData={currentTreeData}
              isTreeQueryError={isTreeQueryError}
            />
          )}
        </>
      ) : (
        noDataChild
      )}
    </TreeDetailsContainer>
  );
}
