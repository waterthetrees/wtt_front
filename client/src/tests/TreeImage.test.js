import React from 'react';
import { render, screen, act } from '@testing-library/react';
import { ImageLoad, fixScientificForImage } from '@/pages/Tree/TreeImage';

describe('ImageLoad', () => {
  const src = 'http://example.com/image.jpg';
  const placeholder = 'http://example.com/placeholder.jpg';
  const alt = 'Sample image';

  it('Renders the ImageLoad component correctly', () => {
    const { container } = render(
      <ImageLoad src={src} placeholder={placeholder} alt={alt} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('loads the actual image and updates the src', async () => {
    const img = new Image();
    const loadEvent = new Event('load');

    // Mock Image constructor to simulate image loading
    jest.spyOn(global, 'Image').mockImplementation(() => {
      img.src = src;
      setTimeout(() => img.dispatchEvent(loadEvent), 50);
      return img;
    });

    render(<ImageLoad src={src} placeholder={placeholder} alt={alt} />);

    await act(async () => new Promise((resolve) => setTimeout(resolve, 60)));

    const renderedImg = screen.getByRole('img', { name: alt });
    expect(renderedImg.getAttribute('src')).toEqual(src);

    // Cleanup
    global.Image.mockRestore();
  });
});

describe('fixScientificForImage', () => {
  it('returns the correct image path when the scientific name is provided and exists in treeImages', () => {
    const scientificName = 'Quercus alba';
    const expectedImagePath = '../../assets/images/data/Quercus-alba.jpg';
    const result = fixScientificForImage(scientificName);

    expect(result).toBe(expectedImagePath);
  });

  it('returns null when the scientific name is not provided', () => {
    const result = fixScientificForImage(null);
    expect(result).toBe(null);
  });

  it('returns null when the scientific name does not exist in treeImages', () => {
    const scientificName = 'Nonexistent scientific name';
    const result = fixScientificForImage(scientificName);
    expect(result).toBe(null);
  });
});
