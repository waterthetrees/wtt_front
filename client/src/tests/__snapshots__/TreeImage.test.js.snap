// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageLoad Renders the ImageLoad component correctly 1`] = `
<div>
  <img
    alt="Sample image"
    loading="lazy"
    placeholder="http://example.com/placeholder.jpg"
    style="opacity: 0.1; transition: opacity .30s linear; width: 100%;"
  />
</div>
`;
